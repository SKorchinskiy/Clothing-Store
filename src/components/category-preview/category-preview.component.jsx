import {
  CategoryPreviewItems,
  CategoryPreviewContainer,
} from "./category-preview.styles";

import { Fragment } from "react";
import { useNavigate } from "react-router-dom";

import ProductItem from "../product-item/product-item.component";

function CategoryPreview({ category }) {
  const { title, items } = category;

  const navigate = useNavigate();

  const goToCategory = () => navigate(`/category/${title.toLowerCase()}`);

  return (
    <CategoryPreviewContainer>
      <h2 onClick={goToCategory}>{title}</h2>
      <CategoryPreviewItems>
        {items.map((item) => (
          <Fragment key={item.id}>
            <ProductItem product={item}></ProductItem>
          </Fragment>
        ))}
      </CategoryPreviewItems>
    </CategoryPreviewContainer>
  );
}

export default CategoryPreview;
