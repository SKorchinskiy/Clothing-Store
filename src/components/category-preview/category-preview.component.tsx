import {
  CategoryPreviewItems,
  CategoryPreviewContainer,
} from "./category-preview.styles";

import { Fragment } from "react";
import { useNavigate } from "react-router-dom";

import ProductItem, {
  ProductType,
} from "../product-item/product-item.component";

type CategoryPreviewType = {
  id: number;
  title: string;
  items: Array<ProductType>;
};

type CategoryPreviewInputType = {
  category: CategoryPreviewType;
};

function CategoryPreview({ category }: CategoryPreviewInputType) {
  const { title, items } = category;

  const navigate = useNavigate();

  const goToCategory = () => navigate(`/category/${title.toLowerCase()}`);

  return (
    <CategoryPreviewContainer>
      <h2>
        <span onClick={goToCategory}>{title.toUpperCase()}</span>
      </h2>
      <CategoryPreviewItems>
        {items.map((item) => (
          <Fragment key={item.id}>
            <ProductItem product={item}></ProductItem>
          </Fragment>
        ))}
      </CategoryPreviewItems>
    </CategoryPreviewContainer>
  );
}

export default CategoryPreview;
