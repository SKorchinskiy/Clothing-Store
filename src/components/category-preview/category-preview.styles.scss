.category-preview-container {
  display: flex;
  flex-direction: column;

  h2 {
    margin-bottom: 20px;
  }

  .category-preview-items {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
    column-gap: 20px;
    row-gap: 30px;
  }
}
