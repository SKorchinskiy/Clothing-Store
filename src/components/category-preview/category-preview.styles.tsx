import styled from "styled-components";

export const CategoryPreviewItems = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
  column-gap: 20px;
  row-gap: 30px;

  @media screen and (max-width: 400px) {
    width: 80%;
    margin: auto;
  }

  @media screen and (min-width: 1100px) {
    display: flex;
    flex-wrap: wrap;
  }
`;

export const CategoryPreviewContainer = styled.div`
  display: flex;
  flex-direction: column;

  h2 {
    margin-bottom: 20px;
    span {
      cursor: pointer;
      font-family: "Josefin Slab";
    }
  }
`;
