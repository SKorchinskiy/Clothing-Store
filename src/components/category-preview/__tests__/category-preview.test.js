import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import { act, screen } from "@testing-library/react";

import CategoryPreview from "../category-preview.component";

import { categoryPreview } from "./stubs/category-preview.stub";

import renderWithProviders from "../../../utils/test.utils";

const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => {
  const actualModule = jest.requireActual("react-router-dom");
  return {
    ...actualModule,
    useNavigate: () => mockedUseNavigate,
  };
});

describe("<CategoryPreview /> component", () => {
  it("should render initial UI properly", () => {
    const { asFragment } = renderWithProviders(
      <CategoryPreview category={categoryPreview} />
    );

    const titleElement = screen.getByText(
      new RegExp(categoryPreview.title, "i")
    );
    expect(titleElement).toBeInTheDocument();

    const imageElements = categoryPreview.items.map((item) =>
      screen.getAllByAltText(item.name)
    );
    expect(imageElements).toHaveLength(3);

    const nameElements = categoryPreview.items.map((item) =>
      screen.getByText(item.name)
    );
    expect(nameElements).toHaveLength(3);

    const priceElements = categoryPreview.items.map((item) =>
      screen.getByText(`$${item.price}`)
    );
    expect(priceElements).toHaveLength(3);

    expect(asFragment()).toMatchSnapshot();
  });

  it("shuold navigate to specific category page", async () => {
    renderWithProviders(<CategoryPreview category={categoryPreview} />);

    const titleElement = screen.getByText(
      new RegExp(categoryPreview.title, "i")
    );

    await act(async () => await userEvent.click(titleElement));

    expect(mockedUseNavigate).toHaveBeenCalledWith(
      `/category/${categoryPreview.title.toLowerCase()}`
    );
  });
});
