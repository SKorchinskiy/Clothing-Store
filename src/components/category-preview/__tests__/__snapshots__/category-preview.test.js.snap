// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryPreview /> component should render initial UI properly 1`] = `
<DocumentFragment>
  <div
    class="sc-grXZZQ dJIsEv"
  >
    <h2>
      <span>
        CATEGORY/TITLE/1
      </span>
    </h2>
    <div
      class="sc-bcPKhP jhARyy"
    >
      <div
        class="sc-hLclGa bXBury"
      >
        <img
          alt="category/1/item/name/1"
          src="category/1/item/imageUrl/1"
        />
        <div
          class="sc-dkjKgF XfSlv"
        >
          <span>
            category/1/item/name/1
          </span>
          <span>
            $category/1/item/price/1
          </span>
        </div>
        <button
          class="sc-eDnVMP sc-gKHVLF krXTxx rLZXw"
          type="button"
        >
          Add To Cart
        </button>
      </div>
      <div
        class="sc-hLclGa bXBury"
      >
        <img
          alt="category/1/item/name/2"
          src="category/1/item/imageUrl/2"
        />
        <div
          class="sc-dkjKgF XfSlv"
        >
          <span>
            category/1/item/name/2
          </span>
          <span>
            $category/1/item/price/2
          </span>
        </div>
        <button
          class="sc-eDnVMP sc-gKHVLF krXTxx rLZXw"
          type="button"
        >
          Add To Cart
        </button>
      </div>
      <div
        class="sc-hLclGa bXBury"
      >
        <img
          alt="category/1/item/name/3"
          src="category/1/item/imageUrl/3"
        />
        <div
          class="sc-dkjKgF XfSlv"
        >
          <span>
            category/1/item/name/3
          </span>
          <span>
            $category/1/item/price/3
          </span>
        </div>
        <button
          class="sc-eDnVMP sc-gKHVLF krXTxx rLZXw"
          type="button"
        >
          Add To Cart
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
